File: backend/internal/service/doctors_02.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// DoctorRecord02Service encapsulates business rules for DoctorRecord02.
type DoctorRecord02Service struct {
	repo *repository.DoctorRecord02Repository
}

// NewDoctorRecord02Service wires the service.
func NewDoctorRecord02Service(repo *repository.DoctorRecord02Repository) *DoctorRecord02Service {
	return &DoctorRecord02Service{repo: repo}
}

// Create validates and persists a new DoctorRecord02.
func (s *DoctorRecord02Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.DoctorRecord02, error) {
	now := time.Now().UTC()
	rec := &models.DoctorRecord02{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a DoctorRecord02 by id.
func (s *DoctorRecord02Service) Get(ctx context.Context, id string) (*models.DoctorRecord02, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped DoctorRecord02 pages.
func (s *DoctorRecord02Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.DoctorRecord02, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing DoctorRecord02.
func (s *DoctorRecord02Service) Update(ctx context.Context, actor string, rec *models.DoctorRecord02) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a DoctorRecord02.
func (s *DoctorRecord02Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds DoctorRecord02 by external code prefix.
func (s *DoctorRecord02Service) Search(ctx context.Context, facilityID, prefix string) ([]models.DoctorRecord02, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *DoctorRecord02Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/doctors02.ts
/** Domain types for doctors / DoctorRecord02 */
export type DoctorRecord02Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface DoctorRecord02 {
  id: string;
  externalCode: string;
  displayName: string;
  status: DoctorRecord02Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface DoctorRecord02CreateRequest {
  externalCode: string;
  displayName: string;
  status: DoctorRecord02Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface DoctorRecord02ListResponse {
  items: DoctorRecord02[];
  total?: number;
}

export interface DoctorRecord02StatsResponse {
  activeCount: number;
}

export function isDoctorRecord02Active(rec: DoctorRecord02): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatDoctorRecord02Label(rec: DoctorRecord02): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareDoctorRecord02ByPriority(a: DoctorRecord02, b: DoctorRecord02): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
